From a44734df6c24fadbdb001f051cc5580c467caf7d Mon Sep 17 00:00:00 2001
From: Per Bilse <per.bilse@citrix.com>
Date: Wed, 7 Dec 2022 12:17:30 +0100
Subject: ioreq_broadcast(): accept partial broadcast success

Avoid incorrectly triggering an error when a broadcast buffered ioreq
is not handled by all registered clients, as long as the failure is
strictly because the client doesn't handle buffered ioreqs.

Signed-off-by: Per Bilse <per.bilse@citrix.com>
Reviewed-by: Paul Durrant <paul@xen.org>

diff --git a/xen/arch/x86/hvm/ioreq.c b/xen/arch/x86/hvm/ioreq.c
index b8e7c2d08929..adc79b460f4d 100644
--- a/xen/arch/x86/hvm/ioreq.c
+++ b/xen/arch/x86/hvm/ioreq.c
@@ -1504,7 +1504,8 @@ unsigned int hvm_broadcast_ioreq(ioreq_t *p, bool buffered)
 
     FOR_EACH_IOREQ_SERVER(d, id, s)
     {
-        if ( !s->enabled )
+        if ( !s->enabled ||
+             (buffered && s->bufioreq_handling == HVM_IOREQSRV_BUFIOREQ_OFF) )
             continue;
 
         if ( hvm_send_ioreq(s, p, buffered) == X86EMUL_UNHANDLEABLE )
